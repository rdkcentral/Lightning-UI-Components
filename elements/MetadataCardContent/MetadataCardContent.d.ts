import type MetadataBase from '../MetadataBase/MetadataBase';
import type Provider from '../Provider/Provider';
import { ProviderStyles } from '../Provider/Provider';
import { __TextType } from '@lightningjs/core';

export interface MetadataCardContentStyles {
  titleTextProperties?: __TextType;
  descriptionProperties?: __TextType;
  detailsTextProperties?: __TextType;
  providerStyle?: ProviderStyles;
}

export declare class MetadataCardContent extends MetadataBase {
  details?: string;
  provider?: Provider;
  style?: MetadataCardContentStyles;
}
