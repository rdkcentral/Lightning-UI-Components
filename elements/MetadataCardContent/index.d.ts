import type MetadataBase from '../MetadataBase';
import { MetadataBaseStyles } from '../MetadataBase';
import type Provider from '../Provider';
import { ProviderStyles } from '../Provider';
import { __TextType } from '@lightningjs/core';

export interface MetadataCardContentStyles extends MetadataBaseStyles {
  titleTextProperties?: __TextType;
  descriptionProperties?: __TextType;
  detailsTextProperties?: __TextType;
  providerStyle?: ProviderStyles;
}

export declare class MetadataCardContent extends MetadataBase {
  details?: string;
  provider?: Provider;
  style: MetadataCardContentStyles;
}
