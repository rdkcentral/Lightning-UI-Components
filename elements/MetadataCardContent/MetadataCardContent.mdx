import { Canvas, Story } from '@storybook/addon-docs';
import MetadataCardContent from '.';
import Provider from '../Provider';
import xfinityProviderLogo from '../../assets/images/Xfinity-Provider-Logo-Square.png';

# MetadataCardContent

A component to format metadata on CardContent. Contains a [Provider](?path=/docs/elements-provider--basic) to format [Icon(s)](?path=/docs/elements-icon--png).

## Usage

```js
import { MetadataCardContent } from '@lightning/ui';

class Basic extends lng.Component {
  static _template() {
    return {
      MetadataCardContent: {
        type: MetadataCardContent,
        w: 600,
        h: 250,
        title: 'Title',
        description: 'Description',
        details: 'Details',
        provider: {
          visibleCount: args.visibleCount,
          providers: Array.apply(null, { length: 10 }).map(
            () => xfinityProviderLogoSquare
          )
        }
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-metadatacardcontent--basic" />
</Canvas>

## API

### Properties

| name        | type                                             | required | default | description                                                                          |
| ----------- | ------------------------------------------------ | -------- | ------- | ------------------------------------------------------------------------------------ |
| title       | string                                           | false    | -       | title text                                                                           |
| description | string                                           | false    | -       | description text directly below title                                                |
| details     | string                                           | false    | -       | details text at bottom left of component                                             |
| provider    | [Provider](?path=/docs/elements-provider--basic) | false    | -       | an object of [Provider](?path=/docs/elements-provider--basic) properties to patch in |

### Style Properties

| name     | type   | description                                                                            |
| -------- | ------ | -------------------------------------------------------------------------------------- |
| provider | object | style properties to pass along to the [Provider](?path=/docs/elements-provider--basic) |

### Methods
