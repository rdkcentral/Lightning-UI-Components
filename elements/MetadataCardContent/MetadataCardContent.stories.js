import lng from '@lightningjs/core';
import MetadataCardContent from '.';
import mdx from './MetadataCardContent.mdx';
import xfinityProviderLogoSquare from '../../assets/images/Xfinity-Provider-Logo-Square.png';

export default {
  title: 'Elements / Metadata',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const CardContent = () =>
  class CardContent extends lng.Component {
    static _template() {
      return {
        MetadataCardContent: {
          type: MetadataCardContent
        }
      };
    }
  };
CardContent.argTypes = {
  w: {
    defaultValue: 600,
    table: {
      defaultValue: { summary: 600 }
    },
    control: 'number',
    description: 'width of component'
  },
  h: {
    defaultValue: 250,
    table: {
      defaultValue: { summary: 250 }
    },
    control: 'number',
    description: 'height of component'
  },
  title: {
    defaultValue: 'Title',
    table: {
      defaultValue: { summary: 'Title' }
    },
    control: 'text',
    description: 'title text'
  },
  description: {
    defaultValue: 'Description',
    table: {
      defaultValue: { summary: 'Description' }
    },
    control: 'text',
    description: 'description text'
  },
  details: {
    defaultValue: 'Details',
    table: {
      defaultValue: { summary: 'Details' }
    },
    control: 'text',
    description: 'details text'
  },
  visibleCount: {
    control: { type: 'range', min: 1, max: 10, step: 1 },
    defaultValue: 3,
    description: 'number of visible providers',
    table: {
      defaultValue: { summary: 3 }
    }
  }
};

CardContent.parameters = {
  argActions: {
    visibleCount(visibleCount, component) {
      component.tag('MetadataCardContent').provider = {
        visibleCount: visibleCount,
        providers: Array(10).fill(xfinityProviderLogoSquare)
      };
    }
  }
};
