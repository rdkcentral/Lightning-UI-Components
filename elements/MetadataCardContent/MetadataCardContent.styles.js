import { processThemeStyles } from '../../mixins';

export const variantDefault = 'neutral';

export function base(theme) {
  return {
    detailsTextProperties: theme.typography.body3,
    providerStyle: {
      fadeWidth: theme.spacing(8),
      itemSize: theme.spacing(6)
    }
  };
}

export function variants(theme) {
  return {
    neutral: {
      detailsTextProperties: {
        textColor: theme.colors.coreNeutral
      }
    },
    inverse: {
      detailsTextProperties: {
        textColor: theme.colors.coreInverse
      }
    },
    brand: {
      detailsTextProperties: {
        textColor: theme.colors.coreNeutral
      }
    }
  };
}

export default (theme, variant = variantDefault) =>
  processThemeStyles(theme, variant, base, variants);
