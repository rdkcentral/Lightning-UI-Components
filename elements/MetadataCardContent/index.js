import MetadataBase from '../MetadataBase';
import Provider from '../Provider';
import TextBox from '../TextBox';
import * as styles from './MetadataCardContent.styles.js';
import { withExtensions } from '../../mixins';
import { FadeShader } from '../../textures';

class MetadataCardContent extends MetadataBase {
  static get __componentName() {
    return 'MetadataCardContent';
  }

  static get __themeStyles() {
    return styles;
  }

  static _template() {
    return {
      Text: {
        flex: { direction: 'column', justifyContent: 'flex-start' },
        Title: {
          type: TextBox
        },
        Description: {
          type: TextBox
        }
      },
      DetailsWrapper: {
        DetailsFader: {
          Details: {
            mountY: 0.5,
            type: TextBox
          }
        },
        Provider: {
          mount: 1,
          type: Provider
        }
      }
    };
  }

  static get properties() {
    return ['description', 'details', 'provider', 'title'];
  }

  static get tags() {
    return [
      'Text',
      {
        name: 'Title',
        path: 'Text.Title'
      },
      {
        name: 'Description',
        path: 'Text.Description'
      },
      'DetailsWrapper',
      {
        name: 'DetailsFader',
        path: 'DetailsWrapper.DetailsFader'
      },
      {
        name: 'Details',
        path: 'DetailsWrapper.DetailsFader.Details'
      },
      {
        name: 'Provider',
        path: 'DetailsWrapper.Provider'
      }
    ];
  }

  _update() {
    this._updateLines();
    this._updateProvider();
    this._updatePositions();
  }

  _updateLines() {
    this._Text.w = this.w;
    this._updateTitle();
    this._updateDescription();
    this._updateDetails();
  }

  _updateDescription() {
    this._Description.patch({
      content: this.description,
      textStyle: this.style.descriptionTextProperties,
      wordWrap: true,
      wordWrapWidth: this._Text.w
    });
  }

  _updateDetails() {
    const maxWidth = this._detailsMaxW();
    this._Details.patch({
      content: this.details,
      textStyle: this.style.detailsTextProperties,
      wordWrap: true,
      maxLines: 1,
      wordWrapWidth: maxWidth + this.style.fadeWidth / 2,
      maxLinesSuffix: ''
    });
    if (this._Details.finalW > maxWidth) {
      this._DetailsFader.patch({
        w: maxWidth + this.style.fadeWidth / 2,
        shader: {
          type: FadeShader,
          positionLeft: 0,
          positionRight: this.style.fadeWidth
        }
      });
    } else {
      this._DetailsFader.shader = undefined;
    }
  }

  _updateProvider() {
    if (this.provider) {
      this._Provider.patch({
        ...this.provider,
        style: this.style.providerStyle
      });
    }
  }

  _updatePositions() {
    this._Text.h = this._textH();
    this._Text.w = this._textW();

    this._DetailsWrapper.w = this._textW();
    this._DetailsWrapper.h = this._providerH();
    this._DetailsWrapper.y = this.h - this._DetailsWrapper.h;

    this._Details.y = this._DetailsWrapper.h / 2;

    this._Provider.x = this._DetailsWrapper.w - this._providerW();
    this._Provider.y = this._DetailsWrapper.h - this._providerH();
  }

  _textW() {
    return this.w;
  }

  _textH() {
    return this.h - this._providerH();
  }

  _providerW() {
    return this._Provider._Row._Items.w;
  }

  _providerH() {
    return this._Provider._Row._Items.h;
  }

  _detailsMaxW() {
    return this.w - this._providerW() - this.style.fadeWidth / 2;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }
    return [
      this.title,
      this.description,
      this.details,
      this._Provider.announce
    ];
  }

  set announce(announce) {
    super.announce = announce;
  }
}

export default withExtensions(MetadataCardContent);
