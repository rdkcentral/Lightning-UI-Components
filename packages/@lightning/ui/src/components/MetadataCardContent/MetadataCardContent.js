import {
  withExtensions,
  MetadataBase,
  TextBox,
  FadeShader
} from '@lightning/ui-core';
import Provider from '../Provider';
import * as styles from './MetadataCardContent.styles.js';

class MetadataCardContent extends MetadataBase {
  static get __componentName() {
    return 'MetadataCardContent';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      Text: {
        flex: { direction: 'column', justifyContent: 'flex-start' },
        Title: {
          type: TextBox,
          signals: {
            textBoxChanged: '_resolveTitle'
          }
        },
        Description: {
          type: TextBox,
          signals: {
            textBoxChanged: '_resolveDescription'
          }
        }
      },
      DetailsWrapper: {
        DetailsFader: {
          Details: {
            mountY: 0.5,
            type: TextBox,
            signals: {
              textBoxChanged: '_resolveDetails'
            }
          }
        },
        Provider: {
          mount: 1,
          type: Provider,
          signals: {
            providerChanged: '_resolveProvider'
          }
        }
      }
    };
  }

  static get properties() {
    return ['description', 'details', 'provider', 'title'];
  }

  static get tags() {
    return [
      'Text',
      {
        name: 'Title',
        path: 'Text.Title'
      },
      {
        name: 'Description',
        path: 'Text.Description'
      },
      'DetailsWrapper',
      {
        name: 'DetailsFader',
        path: 'DetailsWrapper.DetailsFader'
      },
      {
        name: 'Details',
        path: 'DetailsWrapper.DetailsFader.Details'
      },
      {
        name: 'Provider',
        path: 'DetailsWrapper.Provider'
      }
    ];
  }

  _setDetails(details) {
    if (details) {
      this._detailsPromise = new Promise(resolve => {
        this._detailsPromiseResolver = resolve;
      });
    } else {
      this._detailsPromise = undefined;
    }
    return details;
  }

  _setProvider(provider) {
    if (provider) {
      this._providerPromise = new Promise(resolve => {
        this._providerPromiseResolver = resolve;
      });
    } else {
      this._providerPromise = undefined;
    }
    return provider;
  }

  async _update() {
    this._updateLines();
    this._updateProvider();
    await Promise.all(
      [
        this._titlePromise,
        this._descriptionPromise,
        this._detailsPromise,
        this._providerPromise
      ].filter(Boolean)
    );
    this._updatePositions();
  }

  _updateLines() {
    this._Text.w = this.w;
    this._updateTitle();
    this._updateDescription();
    this._updateDetails();
  }

  _updateDescription() {
    this._Description.patch({
      content: this.description,
      style: {
        textStyle: {
          ...this.style.descriptionTextStyle,
          wordWrap: true,
          wordWrapWidth: this._Text.w
        }
      }
    });
  }

  _updateDetails() {
    const maxWidth = this._detailsMaxW;
    this._Details.patch({
      content: this.details,
      style: {
        textStyle: {
          ...this.style.detailsTextStyle,
          wordWrap: true,
          maxLines: 1, // sets default for details
          wordWrapWidth: maxWidth + this.style.fadeWidth / 2
        }
      }
    });
    if (this._Details.finalW > maxWidth) {
      this._DetailsFader.patch({
        w: maxWidth + this.style.fadeWidth / 2,
        shader: {
          type: FadeShader,
          positionLeft: 0,
          positionRight: this.style.fadeWidth
        }
      });
    } else {
      this._DetailsFader.shader = undefined;
    }
  }

  _resolveDetails() {
    this._detailsPromiseResolver && this._detailsPromiseResolver();
    this._updatePositions();
  }

  _updateProvider() {
    if (this.provider) {
      this._Provider.patch({
        mode: this.mode,
        ...this.provider,
        style: this.style.providerStyle
      });
    }
  }

  _resolveProvider() {
    this._providerPromiseResolver && this._providerPromiseResolver();
    this._updatePositions();
  }

  _updatePositions() {
    this._Text.h = this._textH;
    this._Text.w = this._textW;

    this._DetailsWrapper.w = this._textW;
    this._DetailsWrapper.h = Math.max(this._providerH, this._Details.h);
    this._DetailsWrapper.y = this.h - this._DetailsWrapper.h;

    this._Details.y = this._DetailsWrapper.h / 2;

    this._Provider.x = this._DetailsWrapper.w - this._providerW;
    this._Provider.y = this._DetailsWrapper.h - this._providerH;
  }

  get _textW() {
    return this.w;
  }

  get _textH() {
    return this.h - this._providerH;
  }

  get _providerW() {
    return this._Provider.w;
  }

  get _providerH() {
    return this._Provider._Row._Items.h;
  }

  get _detailsMaxW() {
    return this.w - this._providerW - this.style.fadeWidth / 2;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Title && this._Title.announce,
        this._Description && this._Description.announce,
        this._Details && this._Details.announce,
        this._Provider && this._Provider.announce
      ]
    );
  }
}

export default withExtensions(MetadataCardContent);
