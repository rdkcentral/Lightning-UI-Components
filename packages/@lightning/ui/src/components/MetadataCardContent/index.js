import MetadataCardContent from './MetadataCardContent';

export { MetadataCardContent as default };
