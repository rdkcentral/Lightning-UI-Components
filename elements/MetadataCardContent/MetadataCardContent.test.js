import TestUtils from '../../test/lightning-test-utils';
import MetadataCardContent from '.';
import xfinityProviderLogoSquare from '../../assets/images/Xfinity-Provider-Logo-Square.png';

const createComponent = TestUtils.makeCreateComponent(MetadataCardContent);

describe('MetadataCardContent', () => {
  let component, testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createComponent(
      {
        provider: {
          visibleCount: 3,
          providers: Array(10).fill(xfinityProviderLogoSquare)
        }
      },
      {
        spyOnMethods: ['_update']
      }
    );
  });
  afterEach(() => {
    component = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('updates the title', async () => {
    const title = 'title text';
    expect(component.title).toBe(undefined);
    component.title = title;
    await component.__updateSpyPromise;
    expect(component._Title.content).toBe(title);
  });

  it('updates the description', async () => {
    const description = 'description text';
    expect(component.description).toBe(undefined);
    component.description = description;
    await component.__updateSpyPromise;
    expect(component._Description.content).toBe(description);
  });

  it('updates the details', async () => {
    const details = 'details text';
    expect(component.details).toBe(undefined);
    component.details = details;
    await component.__updateSpyPromise;
    expect(component._Details.content).toBe(details);
  });

  it('creates provider', () => {
    expect(component._Provider._Row.items[3].tag('Text').content).toBe('+7');
    expect(component._Provider._Row._Items.h).toBe(
      component.style.providerStyle.itemSize
    );
  });
});
