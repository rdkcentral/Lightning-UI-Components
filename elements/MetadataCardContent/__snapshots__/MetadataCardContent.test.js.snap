// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataCardContent renders 1`] = `
Object {
  "Component": Object {
    "DetailsWrapper": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "DetailsFader": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Details": Object {
              "active": false,
              "alpha": 0.001,
              "attached": true,
              "boundsMargin": null,
              "children": Object {
                "Text": Object {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Text",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": Array [
                    "Text",
                  ],
                  "texture": Object {
                    "textBaseline": "bottom",
                    "type": "TextTexture",
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Details",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": Array [
                "Details",
              ],
              "type": "TextBox",
              "visible": false,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "DetailsFader",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "DetailsFader",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Provider": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Row": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": Object {
                "Items": Object {
                  "active": false,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "children": Object {
                    "Element-15": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": true,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 48,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": true,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 48,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-16": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": true,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 48,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": true,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 48,
                      "x": 56,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-17": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": true,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 48,
                      "hasFinalFocus": false,
                      "hasFocus": false,
                      "isComponent": true,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": true,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": "",
                      "tag": [Function],
                      "type": "Icon",
                      "visible": true,
                      "w": 48,
                      "x": 112,
                      "y": 0,
                      "zIndex": 0,
                    },
                    "Element-18": Object {
                      "active": false,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "children": Object {
                        "Background": Object {
                          "active": false,
                          "alpha": 1,
                          "attached": true,
                          "boundsMargin": null,
                          "clipping": false,
                          "color": 4294967295,
                          "enabled": true,
                          "flex": false,
                          "flexItem": false,
                          "h": 48,
                          "isComponent": undefined,
                          "mount": 0,
                          "mountX": 0,
                          "mountY": 0,
                          "pivot": 0.5,
                          "pivotX": 0.5,
                          "pivotY": 0.5,
                          "ref": "Background",
                          "renderOfScreen": undefined,
                          "renderToTexture": false,
                          "scale": 1,
                          "scaleX": 1,
                          "scaleY": 1,
                          "state": undefined,
                          "tag": [Function],
                          "tags": Array [
                            "Background",
                          ],
                          "texture": Object {
                            "type": "StaticCanvasTexture",
                          },
                          "visible": true,
                          "w": 48,
                          "x": 0,
                          "y": 0,
                          "zIndex": 0,
                        },
                        "Text": Object {
                          "active": false,
                          "alpha": 0.001,
                          "attached": true,
                          "boundsMargin": null,
                          "children": Object {
                            "Text": Object {
                              "active": false,
                              "alpha": 1,
                              "attached": true,
                              "boundsMargin": null,
                              "clipping": false,
                              "color": 4294967295,
                              "enabled": true,
                              "flex": false,
                              "flexItem": false,
                              "h": 0,
                              "isComponent": undefined,
                              "mount": 0,
                              "mountX": 0,
                              "mountY": 0,
                              "pivot": 0.5,
                              "pivotX": 0.5,
                              "pivotY": 0.5,
                              "ref": "Text",
                              "renderOfScreen": undefined,
                              "renderToTexture": false,
                              "scale": 1,
                              "scaleX": 1,
                              "scaleY": 1,
                              "state": undefined,
                              "tag": [Function],
                              "tags": Array [
                                "Text",
                              ],
                              "texture": Object {
                                "fontFace": "XfinityBrownBold",
                                "fontSize": 28,
                                "letterSpacing": -0.2,
                                "lineHeight": 40,
                                "text": "+7",
                                "textBaseline": "bottom",
                                "type": "TextTexture",
                                "verticalAlign": "middle",
                              },
                              "visible": true,
                              "w": 0,
                              "x": 0,
                              "y": 2,
                              "zIndex": 0,
                            },
                          },
                          "clipping": false,
                          "color": 4294967295,
                          "enabled": true,
                          "flex": false,
                          "flexItem": false,
                          "h": 0,
                          "hasFinalFocus": false,
                          "hasFocus": false,
                          "isComponent": true,
                          "mount": 0.5,
                          "mountX": 0.5,
                          "mountY": 0.5,
                          "pivot": 0.5,
                          "pivotX": 0.5,
                          "pivotY": 0.5,
                          "ref": "Text",
                          "renderOfScreen": undefined,
                          "renderToTexture": false,
                          "scale": 1,
                          "scaleX": 1,
                          "scaleY": 1,
                          "state": "",
                          "tag": [Function],
                          "tags": Array [
                            "Text",
                          ],
                          "type": "TextBox",
                          "visible": true,
                          "w": 0,
                          "x": 24,
                          "y": 24,
                          "zIndex": 0,
                        },
                      },
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 48,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": null,
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "visible": true,
                      "w": 48,
                      "x": 168,
                      "y": 0,
                      "zIndex": 0,
                    },
                  },
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 48,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Items",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": Array [
                    "Items",
                  ],
                  "visible": true,
                  "w": 216,
                  "x": 0,
                  "y": 0,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 48,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Row",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "Row",
              "tag": [Function],
              "tags": Array [
                "Row",
              ],
              "type": "Row",
              "visible": true,
              "w": 1920,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 1,
          "mountX": 1,
          "mountY": 1,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Provider",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Provider",
          ],
          "type": "Provider",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "DetailsWrapper",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "DetailsWrapper",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Text": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Description": Object {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Text": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Text",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Text",
              ],
              "texture": Object {
                "textBaseline": "bottom",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Description",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Description",
          ],
          "type": "TextBox",
          "visible": false,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Title": Object {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Text": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Text",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Text",
              ],
              "texture": Object {
                "textBaseline": "bottom",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Title",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": Array [
            "Title",
          ],
          "type": "TextBox",
          "visible": false,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Text",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Text",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
